### PR TITLE
Add support for ref returning methods and properties in `With` statement.

### LINKED ISSUE #68194: System.InvalidOperationException: Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind'
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1818036

_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/SystemInvalidOperationException:-Unexpe/10360956)._

---
[severity:I'm unable to use this version]
I have a mixed C#/VB.NET solution. In C#, there is a class with ref-indexer (`ref T this[int i]`).
In VB, return value of the indexer is used in a `With` statement (`With mWeeks(WeekIndex)`)
This results in exception in Roslyn: `Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind'` which breaks every active analyser etc.
F.e. if the CS project is referenced from VB, but the class name is not fully qualified, pressing Alt-Enter on the `mWeeks` variable below
results in "Feature 'Missing import discovery' is currently unavailable due to an internal error." with the stack trace below.

VB.NET class:
```vb
Public Class Class1
    Private Structure WeekType
        Dim WeekNumber As Integer
    End Structure

    Private mWeeks As MySpecialList(Of WeekType)

    Sub S()
        Dim WeekIndex As Integer
        With mWeeks(WeekIndex)
            .WeekNumber = 0
        End With

        Dim weekNumber = mWeeks(WeekIndex).WeekNumber
    End Sub
End Class
```

C# class:
```cs
namespace CSProj
{
    public class MySpecialList<T>
    {
        private T dummy;
        public ref T this[int i]
        {
            get
            {
                return ref dummy;
            }
        }
    }
}
```

Stacktrace:
```
StreamJsonRpc.RemoteInvocationException: One or more errors occurred. (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind') (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind') (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind')
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__145`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__22`1.MoveNext()
RPC server exception:
System.AggregateException: One or more errors occurred. (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind') (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind') (Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind')
 ---> System.InvalidOperationException: Unexpected value 'PropertyAccess' of type 'Microsoft.CodeAnalysis.VisualBasic.BoundKind'
      at Microsoft.CodeAnalysis.VisualBasic.WithExpressionRewriter.CaptureWithExpression(BoundExpression value, State state)
      at Microsoft.CodeAnalysis.VisualBasic.WithExpressionRewriter.AnalyzeWithExpression(Symbol containingMember, BoundExpression value, Boolean doNotUseByRefLocal, Binder binder, Boolean preserveIdentityOfLValues)
      at Microsoft.CodeAnalysis.VisualBasic.WithBlockBinder.EnsureExpressionAndPlaceholder()
      at Microsoft.CodeAnalysis.VisualBasic.WithBlockBinder.CreateBoundWithBlock(WithBlockSyntax node, Binder boundBlockBinder, BindingDiagnosticBag diagnostics)
      at Microsoft.CodeAnalysis.VisualBasic.Binder.CreateBoundWithBlock(WithBlockSyntax node, Binder boundBlockBinder, BindingDiagnosticBag diagnostics)
      at Microsoft.CodeAnalysis.VisualBasic.Binder.BindStatement(StatementSyntax node, BindingDiagnosticBag diagnostics)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.IncrementalBinder.BindStatement(StatementSyntax node, BindingDiagnosticBag diagnostics)
      at Microsoft.CodeAnalysis.VisualBasic.VBSemanticModel.Bind(Binder binder, SyntaxNode node, BindingDiagnosticBag diagnostics)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.GuardedIncrementalBind(SyntaxNode bindingRoot, Binder enclosingBinder)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.GetBoundNodes(SyntaxNode node)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.GetUpperBoundNode(SyntaxNode node)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.GetBoundNodeSummary(VisualBasicSyntaxNode node)
      at Microsoft.CodeAnalysis.VisualBasic.MemberSemanticModel.GetExpressionSymbolInfo(ExpressionSyntax node, SymbolInfoOptions options, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.SyntaxTreeSemanticModel.GetExpressionSymbolInfo(ExpressionSyntax node, SymbolInfoOptions options, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.VBSemanticModel.GetSymbolInfo(ExpressionSyntax expression, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.VBSemanticModel.GetSymbolInfoForNode(SyntaxNode node, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.VBSemanticModel.GetSymbolInfoCore(SyntaxNode node, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.ModelExtensions.GetSymbolInfo(SemanticModel semanticModel, SyntaxNode node, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService.Rewriter.GetAliasOrAnySymbol(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService.Rewriter.VisitIdentifier(SyntaxToken token, SyntaxToken newToken)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService.Rewriter.VisitToken(SyntaxToken token)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService._Closure$__6-0._Lambda$__2(SyntaxToken t)
      at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   --- End of stack trace from previous location ---
      at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
      at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
      at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   --- End of inner exception stack trace ---
      at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
      at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   --- End of stack trace from previous location ---
      at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
      at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
      at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
      at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, Action`1 body)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService.AddReplacementsWorker(SemanticModel semanticModel, SyntaxNode root, TextSpan span, ConcurrentDictionary`2 replacements, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.VisualBasic.CaseCorrection.VisualBasicCaseCorrectionService.AddReplacements(SemanticModel semanticModel, SyntaxNode root, ImmutableArray`1 spans, ConcurrentDictionary`2 replacements, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.CaseCorrection.AbstractCaseCorrectionService.CaseCorrect(SemanticModel semanticModel, SyntaxNode root, ImmutableArray`1 spans, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.CaseCorrection.AbstractCaseCorrectionService.CaseCorrectAsync(Document document, ImmutableArray`1 spans, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.CaseCorrection.CaseCorrector.CaseCorrectAsync(Document document, SyntaxAnnotation annotation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.CodeActions.CodeAction.CleanupDocumentAsync(Document document, CodeCleanupOptions options, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.AddImport.AbstractAddImportFeatureService`1.SymbolReference.GetTextChangesAsync(Document document, SyntaxNode contextNode, CodeCleanupOptions options, Boolean hasExistingImport, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.AddImport.AbstractAddImportFeatureService`1.SymbolReference.TryGetFixDataAsync(Document document, SyntaxNode node, CodeCleanupOptions options, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.AddImport.AbstractAddImportFeatureService`1.GetFixesInCurrentProcessAsync(Document document, TextSpan span, String diagnosticId, Int32 maxResults, ISymbolSearchService symbolSearchService, AddImportOptions options, ImmutableArray`1 packageSources, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.AddImport.AbstractAddImportFeatureService`1.GetFixesAsync(Document document, TextSpan span, String diagnosticId, Int32 maxResults, ISymbolSearchService symbolSearchService, AddImportOptions options, ImmutableArray`1 packageSources, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.RemoteMissingImportDiscoveryService.<>c__DisplayClass3_0.<<GetFixesAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.<>c__DisplayClass7_0`1.<<RunWithSolutionAsync>g__ProcessSolutionAsync|1>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.RunWithSolutionAsync[T](AssetProvider assetProvider, Checksum solutionChecksum, Int32 workspaceVersion, Boolean updatePrimaryBranch, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.RunWithSolutionAsync[T](AssetProvider assetProvider, Checksum solutionChecksum, Int32 workspaceVersion, Boolean updatePrimaryBranch, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.RunWithSolutionAsync[T](Checksum solutionChecksum, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.RunServiceImplAsync[T](Func`2 implementation, CancellationToken cancellationToken)
```

---
#### Feedback Bot on 5/11/2023, 06:14 PM: 

(private comment, text removed)
#### Feedback Bot on 5/15/2023, 11:09 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #68194.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenRefReturnTests.vb [hunk 1]
```diff
@@ -2577,6 +2577,443 @@ End Module
...
+        Dim o = New C(Of Integer)()
+        With o.GetP()
+        End With
+    End Sub
+End Module",
+                referencedCompilations:={comp1},
+                compilationOptions:=TestOptions.DebugDll)
+            comp2.AssertTheseDiagnostics(
+<expected>
+BC37326: A call to a method or property that returns by reference may not be used in this context.
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenRefReturnTests.vb:2683]: >this context. [](http://example.com/codeflow?start=85&length=13)

nit: I had a bit of trouble understanding this error until I noticed `async`. Consider adding more details in the error to give a clue about what context is okay or not.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Pull request merged**